lsp-mode.el

```diff
@@ -2218,27 +2218,34 @@ WORKSPACE is the workspace that contains the diagnostics."
 
 
 ;; keybindings
+(defvar lsp--binding-descriptions nil
+  "List of key binding/short description pair.")
 
-(defmacro lsp-define-conditional-key (keymap key def cond &rest bindings)
+(defmacro lsp-define-conditional-key (keymap key def desc cond &rest bindings)
   "In KEYMAP, define key sequence KEY as DEF conditionally.
 This is like `define-key', except the definition disappears
 whenever COND evaluates to nil.
-BINDINGS is a list of (key def cond)."
+DESC is the short-description for the binding.
+BINDINGS is a list of (key def desc cond)."
   (declare (indent defun)
-           (debug (form form form form &rest sexp)))
-  (->> (cl-list* key def cond bindings)
-       (-partition 3)
-       (-map (-lambda ((key def cond))
-               `(define-key ,keymap ,key
-                  '(menu-item
-                    ,(format "maybe-%s" def)
-                    ,def
-                    :filter (lambda (item)
-                              (when (with-current-buffer (or (when (buffer-live-p lsp--describe-buffer)
-                                                               lsp--describe-buffer)
-                                                             (current-buffer))
-                                      ,cond)
-                                item))))))
+           (debug (form form form form form &rest sexp)))
+  (->> (cl-list* key def desc cond bindings)
+       (-partition 4)
+       (-mapcat (-lambda ((key def desc cond))
+                  `((define-key ,keymap ,key
+                      '(menu-item
+                        ,(format "maybe-%s" def)
+                        ,def
+                        :filter
+                        (lambda (item)
+                          (when (with-current-buffer (or (when (buffer-live-p lsp--describe-buffer)
+                                                           lsp--describe-buffer)
+                                                         (current-buffer))
+                                  ,cond)
+                            item))))
+                    (when (stringp ,key)
+                      (setq lsp--binding-descriptions
+                            (nconc lsp--binding-descriptions '(,key ,desc)))))))
```

Comment:
nit: using cl-list* will look a bit more natural.

Reply:
They are not equivalent since `lsp--binding-descriptions` is list already. What I want here is `(append A B)` instead of `(cons A B)` like in `cl-list*`.

Reply:
Oh, it seems I misunderstand your point perhaps this comment is for the `lsp-enable-which-key-integration`? The `nconc` there is not correct, indeed.